plugins/kubernetes/app/models/kubernetes/deploy_executor.rb

```diff
@@ -88,76 +89,69 @@ def wait_for_resources_to_complete(release, release_docs)
+    def fetch_pods(release)
+      release.clients.flat_map do |client, query|
+        client.get_pods(query).map! do |p|
+          pod = Kubernetes::Api::Pod.new(p)
+          pod.client = client
+          pod
+        end
+      end
     end
 
-    def show_failure_cause(release)
-      bad_pods(release).each do |pod, client, deploy_group|
+    def show_failure_cause(release, release_docs)
+      pod_statuses(release, release_docs).reject(&:live).select(&:pod).each do |status|
+        pod = status.pod
+        deploy_group = deploy_group_for_pod(pod, release)
         @output.puts "\n#{deploy_group.name} pod #{pod.name}:"
-        print_events(client, pod)
+        print_events(pod)
         @output.puts
-        print_logs(client, pod)
+        print_logs(pod)
         @output.puts "\n------------------------------------------\n"
       end
     end
 
-    # logs - container fails to boot
-    def print_logs(client, pod)
+    # show why container failed to boot
+    def print_logs(pod)
       @output.puts "LOGS:"
 
       pod.containers.map(&:name).each do |container|
         @output.puts "Container #{container}" if pod.containers.size > 1
 
-        logs = begin
-          client.get_pod_log(pod.name, pod.namespace, previous: pod.restarted?, container: container)
-        rescue KubeException
-          begin
-            client.get_pod_log(pod.name, pod.namespace, previous: !pod.restarted?, container: container)
-          rescue KubeException
-            "No logs found"
-          end
-        end
         # Display the first and last n_lines of the log
         max = 50
-        lines = logs.split("\n")
+        lines = (pod.logs(container) || "No logs found").split("\n")
         lines = lines.first(max / 2) + ['...'] + lines.last(max / 2) if lines.size > max
         lines.each { |line| @output.puts "  #{line}" }
       end
     end
 
-    def print_events(client, pod)
+    # show what happened in kubernetes internally since we might not have any logs
+    def print_events(pod)
       @output.puts "EVENTS:"
-      events = client.get_events(
-        namespace: pod.namespace,
-        field_selector: "involvedObject.name=#{pod.name}"
-      )
+      events = pod.events
       events.uniq! { |e| e.message.split("\n").sort }
       events.each { |e| @output.puts "  #{e.reason}: #{e.message}" }
     end
 
-    def bad_pods(release)
-      release.clients.flat_map do |client, query, deploy_group|
-        bad_pods = fetch_pods(client, query).select { |p| p.restarted? || !p.live? }
-        bad_pods.map { |p| [p, client, deploy_group] }
-      end
-    end
-
     def unstable!
-      @output.puts "UNSTABLE - service is restarting"
+      @output.puts "UNSTABLE"
     end
 
+    # user clicked stop button in UI
     def stopped?
       if @stopped
         @output.puts "STOPPED"
         true
       end
     end
 
+    # TODO: cleanup ... a bit ugly with all these arrays
```

Comment:
If you have 4-tuples of values, we should probably just create an object to encapsulate the data.